lib/auth/webauthncli/fido2.go

```diff
@@ -0,0 +1,557 @@
+//go:build libfido2
+// +build libfido2
+
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webauthncli
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/fxamacker/cbor/v2"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"github.com/keys-pub/go-libfido2"
+
+	wanpb "github.com/gravitational/teleport/api/types/webauthn"
+	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
+	log "github.com/sirupsen/logrus"
+)
+
+// FIDO2PollInterval is the poll interval used to check for new FIDO2 devices.
+var FIDO2PollInterval = 200 * time.Millisecond
+
+// FIDODevice abstracts *libfido2.Device for testing.
+type FIDODevice interface {
+	// Info mirrors libfido2.Device.Info.
+	Info() (*libfido2.DeviceInfo, error)
+
+	// Cancel mirrors libfido2.Device.Cancel.
+	Cancel() error
+
+	// Credentials mirrors libfido2.Device.Credentials.
+	Credentials(rpID string, pin string) ([]*libfido2.Credential, error)
+
+	// Assertion mirrors libfido2.Device.Assertion.
+	Assertion(
+		rpID string,
+		clientDataHash []byte,
+		credentialIDs [][]byte,
+		pin string,
+		opts *libfido2.AssertionOpts) (*libfido2.Assertion, error)
+}
+
+// fidoDeviceLocations and fidoNewDevice are used to allow testing.
+var fidoDeviceLocations = libfido2.DeviceLocations
+var fidoNewDevice = func(path string) (FIDODevice, error) {
+	return libfido2.NewDevice(path)
+}
+
+// LoginPrompt is the user interface for FIDO2Login.
+type LoginPrompt interface {
+	// PromptPIN prompts the user for their PIN.
+	PromptPIN() (string, error)
+	// PromptAdditionalTouch prompts the user for an additional security key
+	// touch.
+	// Additional touches may be required after PINs and during passwordless flows.
+	PromptAdditionalTouch() error
+}
+
+// FIDO2Login signs an assertion using available CTAP1 or CTAP2 devices.
+// It must be called with a context with timeout, otherwise it can run
```

Comment:
Hmm. I wonder if there's a way to make this harder to mess up?

Reply:
We can add a hard limit, but this is actually how the U2F code behaves. WDYT is a good hard-limit to add?